src/engine/physics/CharacterPhysicsUpdate.cpp

```diff
@@ -4,41 +4,119 @@ namespace lei3d
         }
         if (glfwGetKey(window, GLFW_KEY_RIGHT) == GLFW_PRESS)
         {
-            velocity = velocity + btVector3(0.0, 0.0, -0.5);
+            wishdir = wishdir + glm::vec3(0.0, 0.0, 1.0);
+        }
+
+        wishdir = glm::normalize(wishdir);
+
+        if (onGround) {
+            glm::vec3 outputVel = GroundAcceleration(wishdir, prevVel);
+            v = btVector3(outputVel.x, outputVel.y, outputVel.z);
+        } else {
+            glm::vec3 outputVel = AirAcceleration(wishdir, prevVel);
+            v = btVector3(outputVel.x, outputVel.y, outputVel.z);
         }
-        if (glfwGetKey(window, GLFW_KEY_SPACE) == GLFW_PRESS)
+
+        if (glfwGetKey(window, GLFW_KEY_SPACE) == GLFW_PRESS && onGround)
         {
-            velocity = velocity + btVector3(0.0, 0.0, 0.5);
+            v = v + btVector3(0.0, 4.75, 0.0);
         }
 
-        m_body->setLinearVelocity(velocity);
+        m_body->setLinearVelocity(v);
     }
 
     void CharacterPhysicsUpdate::debugDraw(btIDebugDraw* debugDrawer) {}
 
-}
+    glm::vec3 CharacterPhysicsUpdate::Accelerate(glm::vec3 wishDir, glm::vec3 prevVel, float acceleration, float maxVelocity)
+    {
+        glm::vec3 outputVelocity = prevVel;
+
+        // project current velocity onto the direction we wish to move
+        glm::vec3 projVel = projectVector(prevVel, wishDir);
+
+        // check if the wish direction is towards or away from the projected velocity
+        bool isAway = glm::dot(wishDir, projVel) <= 0;
+        float projectedVectorMagnitude = glm::length(projVel);
+
+        // only apply force (add speed) if moving away from velocity or velocity is below maxairspeed
+        if (projectedVectorMagnitude < maxVelocity || isAway)
+        {
+            // calculate the ideal movement force
+            glm::vec3 vc = wishDir * acceleration; // NEEDS WISHDIR TO BE NORMALIZED
+            float vcMagnitude = glm::length(vc);
+            // cap the velocity if it would accelerate beyond maxairspeed in direction of velocity
+            if (!isAway) {
+                if (vcMagnitude > maxVelocity - projectedVectorMagnitude) 
+                    vc = (vc / vcMagnitude) * maxVelocity - projectedVectorMagnitude;
+            } else {
+                if (vcMagnitude > maxVelocity + projectedVectorMagnitude) 
+                    vc = (vc / vcMagnitude) * maxVelocity + projectedVectorMagnitude;
+            }
+            outputVelocity = vc;
+        }
+        return outputVelocity;
+    }
+
+    glm::vec3 CharacterPhysicsUpdate::AirAcceleration(glm::vec3 wishDir, glm::vec3 prevVelocity)
+    {
+        return Accelerate(wishDir, prevVelocity, m_airAccel, m_maxAirSpeed);
+    }
+
+    glm::vec3 CharacterPhysicsUpdate::GroundAcceleration(glm::vec3 wishDir, glm::vec3 prevVelocity)
+    {
+        float speed = prevVelocity.length();
+        if (speed != 0)
+        {
+            float drop = speed * m_friction; // THIS MIGHT HAVE TO BE MULTIPLIED BY DELTA TIME??
```

Comment:
Yeah prob., we should maybe store dt in like the application every frame so that it can be accessed without passing it through every function.

Reply:
Good idea, I'll save this for a once I have a working-ish solution and then I'll add this in

Reply:
This (getting dt from the application) seems to already be implemented? Added this functionality as I understand.